### ISSUE TITLE
Transition back to CSS styles

### ISSUE THREAD

**Issue by IPWright83** [mbostock/d3#2396]
I can create a transition that will correctly get applied from a CSS style to an in-line style. For example say I create a circle:

````
var c1 = svg.append("circle")
            .attr("class", "red")
            .attr("r", 25)
            .attr("cx", 50)
            .attr("cy", 50);
````

with some CSS

````
.red { 
   fill: red;  
}
````

I can change the style of this circle to blue using a transition and this works fine.

````
svg.selectAll("circle")
   .transition()
   .delay(2000)
   .duration(2000)
   .style("fill", "blue");
````

What I really want to be able to do, is then revert this style by simply removing it:

````
c1.transition()
  .delay(5000)
  .duration(2000)
  .style("fill", null);
````

This will cause a sudden switch back to red, because I've not defined the colour. What I really would like to be able to do is transition back to the default style from the CSS without having to record the colour red so I can apply it inline to revert back nicely.

Question + Sample on StackOverflow: http://stackoverflow.com/questions/29701689/transitions-styles-back-to-css-defaults#29702008

**Comment by IPWright83** [comment]
The nearest I've got to making this work is to go re-acquire the style of the element. So to go back to red:

````
c1.transition()
  .delay(5000)
  .duration(2000)
  .style("fill", function(d) { return d3.select(this).style("fill"); });
````

This is better than trying to save the colour before making it blue (although I'm kinda surprised this code works). I wonder if this suggests a change to `style` may be useful? Something like (pseudo-ish code):

````
style = function(name, value) {
   if(value === null || value === undefined) { 
     value = d3.select(this).style("name");
  }
  // apply value
}
````

This would make it much nicer to use - although I'm not sure on the performance hit of selecting `this` for each instance changing?

**Comment by mbostock** [comment]
You could also capture the value immediately before starting the transition, like this:

```js
c1.transition()
    .style("fill", function() {
      var that = d3.select(this),
          currentFill = that.style("fill"),
          defaultFill = that.style("fill", null).style("fill"); // retrieve default fill
      that.style("fill", currentFill); // restore current fill
      return defaultFill;
    });
```

Perhaps it would be nice if `transition.style(name, null)` did this for you by default: i.e., computed the style property that you’d get from removing the inline style, and then transitioned to it automatically before removing the inline style on transition end. That seems reasonable.

**Comment by IPWright83** [comment]
@mbostock yeah, I've played around with solutions like that (although not quite so generic) which made my code a little verbose. I think I considered re-defining `style` at some point to add an approach to do this, but it does feel like it'd be a nice addition to core D3.

**Comment by mbostock** [comment]
This has been implemented in d3-transition for 4.0.

**Comment by IPWright83** [comment]
:+1: Just wanted to say thank you @mbostock for looking at this - it helps eliminate more styling information in our JavaScript code, putting it back to its rightful place in CSS land and just makes D3 an even better tool to be working with :)
